sycl/doc/extensions/proposed/sycl_ext_oneapi_forward_progress.asciidoc

```diff
@@ -0,0 +1,611 @@
+scope containing all threads that may coordinate with one another.
+====
+
+=== Forward progress guarantees
+
+The `forward_progress_guarantee` `enum` is used to represent the three classes
+of forward progress guarantee defined in the {cpp} core language specification.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental::sycl {
+
+enum class forward_progress_guarantee {
+  concurrent,
+  parallel,
+  weakly_parallel
+};
+
+}
+----
+
+[NOTE]
+====
+This `enum` is defined similarly to the one defined in
+https://wg21.link/p2300[P2300: `std::execution`].
+====
+
+
+=== Execution scopes
+
+The `execution_scope` `enum` is used to represent the hierarchy of threads of
+execution within a SYCL implementation.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental::sycl {
+
+enum class execution_scope {
+  work_item,
+  sub_group,
+  work_group,
+  root_group,
+};
+
+}
+----
+
+[NOTE]
+====
+The existing `memory_scope` `enum` is not used here to provide greater
+flexibility in modifying the forward progress extension in future.
+`memory_scope` is lacking values required by this extension (i.e. `root_group`)
+and has additional values that are not required here (e.g. `system`).
+====
+
+
+=== Device queries
+
+New device queries are introduced to allow developers to reason about the
+progress guarantees that can be provided by each device.
+
+The results of these queries represent the strongest forward progress
+guarantees that a device can satisfy, and therefore the strongest forward
+progress guarantees that can be requested by a kernel's <<properties,
+properties>>.
+
+The queries are expressed in terms of a _target_ execution scope and a
+_coordination scope_, as described previously. The _coordination_ scope
+must be broader than the _target_ execution scope.
+
+[NOTE]
+====
+Note that these queries cannot be used to reason about the thread of execution
+associated with the root-group, because it is currently the top of the
+hierarchy; when the target execution scope is `root_group`, it is not currently
+possible to specify a broader coordination scope.
+====
+
+[NOTE]
+====
+Explicitly specifying both the target and coordination scope ensures that the
+interpretation of forward progress queries is fixed and independent of the
+number of execution scopes in the hierarchy. This extension therefore does
+not prevent SYCL or other extensions from introducing new execution scopes.
+====
+
+
+[%header,cols="1,5,5"]
+|===
+|Device Descriptor
+|Return Type
+|Description
+
+|`template <execution_scope Scope, execution_scope CoordinationScope>
+ info::device::ext_oneapi_forward_progress_guarantee`
+|`forward_progress_guarantee`
+|Return the strongest forward progress guarantee that this device can support
+ for threads of execution with the specified combination of scope and
+ coordination scope. Specializations of `forward_progress_guarantee` are only
+ available when `CoordinationScope` is broader than `Scope`.
```

Comment:
There's something about this query definition that seems imprecise to me.  There are a couple reasons:

* I think the definition could be more tightly tied to the abstract execution model you present in the "Execution model" section.
* The phrase "with the specified combination of scope and coordination scope" seems a little vague.  Does this imply that the forward progress guarantee of a thread is somehow relative to other threads?  I think this is not the case in the C++ execution model.  In that model each thread has its own independent forward progress guarantee that is independent from the other threads.

I'm also not sure it makes sense for the query to distinguish between `weakly_parallel` vs. `parallel`.  While trying to formalize the query, it was becoming difficult to determine which of these two values to return.  I also wonder if applications care about the difference.  It seems like the only important thing is to tell whether all threads in the coordination scope are guaranteed to have concurrent forward progress.

This makes me wonder if the query should be changed to something like this:

> ```
> template <execution_scope Scope, execution_scope CoordinationScope>
> info::device::ext_oneapi_is_concurrent
> ```
>
> Tells whether all threads of execution of rank _Scope_ within _CoordinationScope_ are guaranteed to make independent forward progress, assuming that at least one thread of execution with rank _Scope_ within _CoordinationScope_ is currently executing.

This definition relies on a new term: the "rank" of a thread of execution.  This ties into the abstract execution model.  The thread that spawns the entire kernel has rank `root_group`.  That thread spawns threads of rank `work_group`.  Those threads spawn threads of rank `sub_group`.  Those threads spawn threads of rank `work_item`.

Let's take your previous hypothetical implementation and consider some example queries:

> For example, consider a hypothetical implementation that creates a separate std::thread (providing concurrent forward progress guarantees) to execute each sub-group, but which permits an unbounded number of work-groups executed as tasks (providing parallel forward progress guarantees). 

Querying `ext_oneapi_is_concurrent<work_item, sub_group>` should return `true`.  The thread executing the sub-group has parallel forward progress.  However, the query assumes that one of the work-items in this sub-group is executing, so we know the sub-group thread has started execution, and we know that it has spawned at least one of the work-item threads in this sub-group.  Since the sub-group thread has parallel forward progress, we know that it will eventually finish spawning all of the other work-items in this sub-group before it blocks.  In addition, each work-item thread is spawned with `std::thread`, which provides concurrent forward progress.  Thus, all threads in this sub-group will eventually be spawned, and they will all be spawned with concurrent forward progress.

Querying `ext_oneapi_is_concurrent<work_item, work_group>` should return `false`.  Each thread in the work-group has parallel forward progress.  We know that some work-item in one of the sub-groups is executing, call this work-group thread A.  However, we don't know whether any of the other work-group threads have started yet.  Since they have only parallel forward progress, there is no guarantee that they will start until after work-group thread A completes.

Querying `ext_oneapi_is_concurrent<sub_group, work_group>` should return `false`.  The logic is similar to the previous case.  Each thread in the work-group has parallel forward progress.  We know that one of these threads (A) has started executing because one of the sub-groups has started.  However, we don't know whether any of the other work-group threads have started yet.  Since they have only parallel forward progress, there is no guarantee that they will start until after work-group thread A completes.

Reply:
I agree that this query could be more tightly tied to the execution model, but I disagree with some of the simplifications you're proposing here.  I think this might end up being quite complicated to work through, so let me know if you'd prefer to jump on a quick call after this.

> The phrase "with the specified combination of scope and coordination scope" seems a little vague.  Does this imply that the forward progress guarantee of a thread is somehow relative to other threads?  I think this is not the case in the C++ execution model.  In that model each thread has its own independent forward progress guarantee that is independent from the other threads.

That wasn't the intent.  I've tried to be very careful to maintain the property that threads only have one forward progress guarantee.  All of the coordination scope stuff is a shorthand for reasoning about the forward progress guarantees at multiple levels of the hierarchy.  Would something like the below be clearer?

```asciidoc
Return the strongest forward progress guarantee that can be passed to the
progress property associated with `Scope`, for the specified value of
`CoordinationScope`.
```

Or if we want to tie it more obviously to the property design we landed on, we could have separate properties for each layer, like an `info::device::work_group_progress<CoordinationScope>` query that returns the strongest forward progress guarantee that can be passed to the `sycl::work_group_progress` property on this device?


> I'm also not sure it makes sense for the query to distinguish between `weakly_parallel` vs. `parallel`. While trying to formalize the query, it was becoming difficult to determine which of these two values to return. I also wonder if applications care about the difference. It seems like the only important thing is to tell whether all threads in the coordination scope are guaranteed to have concurrent forward progress.

I disagree with this.

`weakly_parallel` is the default execution model of SYCL, and it's useful for developers to know if they only have weakly parallel guarantees.  A thread with weakly parallel guarantees can't safely acquire a lock (in general) because you don't know when it will release it.

The difference between `parallel` and `concurrent` is also important, but not for locks.  For locks, both parallel and concurrent threads can safely acquire locks (because we know they'll eventually release the lock).  The difference is important for other coordination routines.  Any thread with `parallel` guarantees has `concurrent` guarantees after executing its first step, so if a thread can prove another thread has taken its first step then it's safe to communicate with it.  For example, `parallel` is sufficient to implement communication between the subset of threads that you can prove have reached a certain point of execution (e.g. because they've all incremented some atomic counter).  `concurrent` provides the additional guarantee that threads _will_ reach a certain point in execution.

My understanding is that there are real algorithms that depend on this difference.  There are implementations of scans that assign tickets to work-groups as they launch (via an atomic), then have later work-groups block on earlier ones. `parallel` forward progress guarantees are sufficient for this.

The alternative queries you outlined are well-defined, but since they don't account for this distinction I think we need a different design.